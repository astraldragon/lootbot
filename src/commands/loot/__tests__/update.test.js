jest.mock("discord-akairo");
jest.mock("../../../models", () => {
  return {
    Loot: { update: jest.fn("update") },
    Tier: { findOne: jest.fn("findOne") }
  };
});

const { Command } = require("discord-akairo");
const { Loot, Tier } = require("../../../models");
const { checkManagePermissions } = require("../../../support");
const UpdateCommand = require("../update");

describe("commands/loot/update", () => {
  beforeEach(() => {
    this.command = new UpdateCommand();
    this.updates = { existingName: "Loot", name: null, tier: null };
    this.tier = { id: 1, name: "Common" };

    this.msg = {
      guild: { id: "1234" },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should have the proper permission check", () => {
    expect(Command.mock.calls[0][1].options.permissions).toBe(
      checkManagePermissions
    );
  });

  it("should update only non-null values", async () => {
    Loot.update.mockReturnValue([1]);
    this.updates = { ...this.updates, name: "New Name" };
    await this.command.exec(this.msg, this.updates);

    expect(Loot.update).toHaveBeenCalledWith(
      { name: this.updates.name },
      { where: { name: this.updates.existingName, guild: this.msg.guild.id } }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.updates.name} updated`
    );
  });

  it("should update the tier if a valid tier is provided", async () => {
    Loot.update.mockReturnValue([1]);
    Tier.findOne.mockReturnValue(this.tier);

    this.updates = { ...this.updates, tier: this.tier.name };
    await this.command.exec(this.msg, this.updates);

    expect(Loot.update).toHaveBeenCalledWith(
      { tier_id: this.tier.id },
      { where: { name: this.updates.existingName, guild: this.msg.guild.id } }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.updates.existingName} updated`
    );
  });

  it("should notify when tier provided does not exist", async () => {
    Tier.findOne.mockReturnValue(null);
    this.updates = { ...this.updates, tier: this.tier.name };
    await this.command.exec(this.msg, this.updates);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `No tier named ${this.updates.tier} found.`
    );
  });

  it("should notify when the loot does not exist", async () => {
    Loot.update.mockReturnValue([0]);
    await this.command.exec(this.msg, this.updates);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.updates.existingName} not found`
    );
  });

  it("should notify if any errors occur", async () => {
    Loot.update.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg, this.updates);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `An error occurred updating ${this.updates.existingName}`
    );
  });
});
