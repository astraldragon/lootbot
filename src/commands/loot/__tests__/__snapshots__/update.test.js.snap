// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/loot/update should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "loot-update",
      Object {
        "aliases": Array [
          "loot-update",
          "lu",
        ],
        "args": Array [
          Object {
            "id": "existingName",
            "prompt": Object {
              "start": "What is the name of the loot?",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "name",
            "match": "prefix",
            "prefix": "name=",
            "prompt": Object {
              "optional": true,
              "start": "What is the new name of the loot?",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "tier",
            "match": "prefix",
            "prefix": "tier=",
            "prompt": Object {
              "optional": true,
              "start": "What is the new tier of the loot?",
            },
            "type": "string",
          },
        ],
        "category": "Loot",
        "channelRestriction": "guild",
        "description": Object {
          "description": "Update that loot",
          "examples": Array [
            "loot-update \\"Maple Syrup\\" name=\\"New Maple Syrup\\" tier=Legendary",
            "loot-update Syrup tier=Legendary",
          ],
          "usage": "<existingName> name=<name> tier=<tier>",
        },
        "options": Object {
          "permissions": [Function],
        },
        "split": "sticky",
      },
    ],
  ],
}
`;
